django/contrib/contenttypes/fields.py

```diff
@@ -224,23 +225,22 @@ def __get__(self, instance, cls=None):
         ct_id = getattr(instance, f.get_attname(), None)
         pk_val = getattr(instance, self.fk_field)
 
-        try:
-            rel_obj = getattr(instance, self.cache_attr)
-        except AttributeError:
-            rel_obj = None
-        else:
-            if rel_obj and (ct_id != self.get_content_type(obj=rel_obj, using=instance._state.db).id or
-                            rel_obj._meta.pk.to_python(pk_val) != rel_obj.pk):
-                rel_obj = None
+        rel_obj = self.get_cached_value(instance, default=None)
 
         if rel_obj is not None:
-            return rel_obj
+            ct_match = ct_id == self.get_content_type(obj=rel_obj, using=instance._state.db).id
+            pk_match = rel_obj._meta.pk.to_python(pk_val) == rel_obj.pk
+
+            if ct_match and pk_match:
+                return rel_obj
+            else:
+                rel_obj = None
 
         if ct_id is not None:
             ct = self.get_content_type(id=ct_id, using=instance._state.db)
             with suppress(ObjectDoesNotExist):
                 rel_obj = ct.get_object_for_this_type(pk=pk_val)
-        setattr(instance, self.cache_attr, rel_obj)
+        self.set_cached_value(instance=instance, value=rel_obj)
```

Comment:
Are you passing args as kwargs like this and throughout the patch because of readability? I'm not sure it helps -- it seems natural that a `set()` method would take `(key, value)`.